doc/design/custom_xcode_schemes.md

```diff
@@ -0,0 +1,467 @@
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+The above example adds a second scheme called `My Application`. It is configured with two test targets
+and a single launch target. All three targets are implicitly included in the build list.
+
+### Specifying Launch Arguments and Environment Variables
+
+There are times when it is desirabe to customize the launch action with arguments and environment
+variables. This can be accomplished by specifying them as parameters to the
+`xcode_schemes.launch_action` function call.
+
+```python
+# Assumptions
+#   //Sources/Foo:Foo - swift_library
+#   //Sources/FooTests:FooTests = ios_unit_test
+#   //Sources/App = ios_application
+#   //Sources/AppUITests = ios_ui_test
+
+load(
+    "@com_github_buildbuddy_io_rules_xcodeproj//xcodeproj:xcodeproj.bzl",
+    "xcode_schemes",
+    "xcodeproj",
+)
+
+_SCHEMES = [
+    xcode_schemes.scheme(
+        name = "Foo Module",
+        build_action = xcode_schemes.build_action(["//Sources/Foo"]),
+        test_action = xcode_schemes.test_action(["//Tests/FooTests"]),
+    ),
+    xcode_schemes.scheme(
+        name = "My Application",
+        test_action = xcode_schemes.test_action([
+            "//Tests/AppUITests",
+            "//Tests/FooTests",
+        ]),
+        launch_action = xcode_schemes.launch_action(
+            target = "//Sources/App",
+            args = [
+                "--my_awesome_flag",
+                "path/to/a/file.txt",
+            ],
+            env = {
+                "RELEASE_THE_KRAKEN": "true",
+            },
+        ),
+    ),
+]
+
+xcodeproj(
+    name = "generate_xcodeproj",
+    project_name = "Command Line",
+    tags = ["manual"],
+    schemes = _SCHEMES,
+)
+```
+
+This example is the same as the previous one except arguments and environment variables are
+defined for the launch action.
+
+## Introduction of Scheme Autogeneration Mode
+
+With this proposal, it is now possible to control how Xcode scheme autogeneration works using
+the `scheme_autogeneration_mode` attribute on `_xcodeproj`. 
+
+__Values__
+- `auto`: If no custom schemes are provided, an Xcode scheme will be created for every buildable
+  target. If custom schemes are provided, no autogenerated schemes will be created.
+- `none`: No custom schemes are generated.
+- `all`: Generate a scheme for every buildable target even if custom schemes are provided.
+
+The default value for `scheme_autogeneration_mode` is `auto`.
+
+## Implementation Changes
+
+### `xcode_schemes` Module
+
+```python
+
+def _scheme(
+    name, 
+    as_json = True, 
+    build_action = None, 
+    test_action = None, 
+    launch_action = None):
+    """Returns a `struct` or JSON `string` representing an Xcode scheme.
+
+    Args:
+        name: The user-visible name for the scheme as a `string`.
+        as_json: Optional. A `bool` indicating whether the resulting struct 
+            should be returned as JSON.
```

Comment:
What's the use case for this?

Reply:
This provides some syntactic sugar for the client. In the implementation, I was planning for all of the functions in `xcode_schemes` to return `struct` values. However, when the client uses `xcode_schemes.scheme`, the attribute on `_xcodeproj` rule will accept a `list` of JSON `string` values. I didn't the client to write something ugly like `xcode_schemes.scheme_json`. So, I added the `as_json` parameter. The client won't touch it, but our implementation and tests would.

As I was writing the above, it occurred to me that we can go at this a slightly different way. Today, we wrap calls to `_xcodeproj` in the macro `xcodeproj`. We could have `xcode_schemes.scheme` return a `struct` like the other functions. The `xcodeproj` macro can then serialize them to JSON before passing them to `_xcodeproj`. This approach has another nice feature. It skips a serialization/deserialization roundtrip. 

What do you think?

Reply:
>The xcodeproj macro can then serialize them to JSON before passing them to _xcodeproj

That is what I was thinking as well. Let's go that route.